common/src/ray_tracing.rs

```diff
@@ -19,39 +32,81 @@ pub fn trace_ray(
-
-    const EPSILON: f32 = 1e-4;
-    let found_collision = if node_ray.direction.z >= 0.0 {
-        // Going the wrong way. Don't report a collision.
-        false
-    } else if (-EPSILON..=0.0).contains(&node_ray.position.z) {
-        // Extra logic to ensure precision issues don't allow a collider to clip through a surface
-        result.tanh_distance = 0.0;
-        true
-    } else {
-        // Compute needed tanh_distance to reach z=0 plane.
-        let tanh_distance = -node_ray.position.z / node_ray.direction.z;
-        if tanh_distance >= 0.0 && tanh_distance < result.tanh_distance {
-            // Report a collision if the needed tanh_distance is in the right range
-            result.tanh_distance = tanh_distance;
-            true
-        } else {
-            false
-        }
-    };
+    let node = graph.get(start_chunk.node).as_ref().unwrap();
+    let Chunk::Populated {
+            voxels: ref voxel_data,
+            ..
+        } = node.chunks[start_chunk.vertex] else {
+            panic!("Collision checking on unpopulated chunk");
+        };
+    let local_ray = start_chunk.vertex.node_to_dual().cast::<f32>() * start_node_transform * ray;
 
-    if found_collision {
-        result.intersection = Some(RayTracingIntersection {
-            normal: math::mtranspose(&start_node_transform) * na::Vector4::z(),
-        });
+    let bounding_box = CubicVoxelRegion::from_ray_segment_and_radius(
+        dimension,
+        &local_ray,
+        result.tanh_distance,
+        chunk_ray_tracer.max_radius(),
+    );
+
+    // Check collision within a single chunk
+    if let Some(bounding_box) = bounding_box {
+        chunk_ray_tracer.trace_ray(
+            &ChunkRayTracingContext {
+                dimension,
+                dimension_f32: dimension as f32,
+                chunk: start_chunk,
+                transform: math::mtranspose(&start_node_transform)
+                    * start_chunk.vertex.dual_to_node().cast(),
+                voxel_data,
+                ray: &local_ray,
+                bounding_box,
+            },
+            &mut result,
+        );
     }
 
     result
 }
 
+/// Wraps any logic needed for a particular collider to perform ray tracing within a given chunk
+pub trait ChunkRayTracer {
+    /// Performs ray tracing with a single chunk. If an intersection is found, `result` is updated
+    /// to reflect this intersection, overriding any old intersections that may have been found.
+    fn trace_ray(&self, ctx: &ChunkRayTracingContext, result: &mut RayTracingResult);
+
+    /// Returns the radius of the sphere of influence of the collider. `trace_ray` might not be called on
+    /// chunks outside this sphere of influence.
+    fn max_radius(&self) -> f32;
+}
+
+/// Contains all the immutable data needed for `ChunkRayTracer` to perform its logic
+pub struct ChunkRayTracingContext<'a> {
+    pub dimension: usize,
+    pub dimension_f32: f32,
```

Comment:
Storing these separately is not necessary. Seems like mostly code wants the `f32`, so let's stick with that.

Reply:
If we want to keep just one of them, I think it makes more sense to keep `dimension`, since it is used in some places, and converting a usize to a float and back to a usize without any other changes seems not ideal.

This context class isn't permanently stored anywhere, so there doesn't seem to be much of a downside to keeping both versions of `dimension`.

That being said, I'm also happy to just use `dimension as f32` everywhere.

Reply:
Sure, sounds good.

> there doesn't seem to be much of a downside to keeping both versions of dimension.

The downside is increased verbosity, decreased readability, and room for error.